### PR TITLE
feat: support merging user and default bootstrap configurations

### PR DESCRIPTION
**Pull request by zhaohuabing**
Related issue: https://github.com/envoyproxy/gateway/issues/1685

Since this is a breaking change, please consider including the following notification, or something similar,  in the release note to remind the users to update their `EnvoyProxy` configuration. 

**Important** The `Booststrap` field of `EnvoyProxy` API has been changed from a string to a struct. Please ensure you adjust your `EnvoyProxy` configurations accordingly. Here is an example:

Before:

```
apiVersion: config.gateway.envoyproxy.io/v1alpha1
kind: EnvoyProxy
metadata:
  name: custom-proxy-config
  namespace: envoy-gateway-system
spec:
  bootstrap: |
    admin:
      access_log: // truncated for brevity
```

After:
```
apiVersion: config.gateway.envoyproxy.io/v1alpha1
kind: EnvoyProxy
metadata:
  name: custom-proxy-config
  namespace: envoy-gateway-system
spec:
  bootstrap:
    type: Replace
    bootstrap: |
      admin:
        access_log: // truncated for brevity
```

### CODE REVIEW

File: api/config/v1alpha1/envoyproxy_types.go [hunk 1]
```diff
@@ -161,6 +161,29 @@ const (
 	LogComponentRuntime LogComponent = "runtime"
 )
 
+// ProxyBootstrap defines Envoy Bootstrap configuration.
+type ProxyBootstrap struct {
+	// Type is the type of the bootstrap configuration, it should be either Merge or Replace.
```
Comment (arkodg) [inline api/config/v1alpha1/envoyproxy_types.go:166]: lets default to `Replace`

File: api/config/v1alpha1/envoyproxy_types.go [hunk 2]
```diff
@@ -161,6 +161,29 @@ const (
 	LogComponentRuntime LogComponent = "runtime"
 )
 
+// ProxyBootstrap defines Envoy Bootstrap configuration.
```
Comment (arkodg) [inline api/config/v1alpha1/envoyproxy_types.go:164]: do we want to handle to list of `merges ?

**Comment by arkodg** [comment]
thanks for picking this up @zhaohuabing, can you split up API and implementation into 2 PRs, which may help with faster reviews, thanks !

**Comment by zhaohuabing** [comment]
> thanks for picking this up @zhaohuabing, can you split up API and implementation into 2 PRs, which may help with faster reviews, thanks !

@arkodg Since this is a breaking change, it won't pass the build if just changing the API.

**Comment by arkodg** [comment]
> > thanks for picking this up @zhaohuabing, can you split up API and implementation into 2 PRs, which may help with faster reviews, thanks !
> 
> @arkodg Since this is a breaking change, it won't pass the build if just changing the API.

ah yah, sure lets stick with this for now

File: api/config/v1alpha1/envoyproxy_types.go (hunk 2 above)
Comment (zhaohuabing) [reply api/config/v1alpha1/envoyproxy_types.go:164]:
```
        // Merge merges the provided bootstrap with the default one. The provided bootstrap can add or override a value
	// within a map, or add a new value to a list.
	// Please note that the provided bootstrap can't override a value within a list.
	BootstrapTypeMerge BootstrapType = "Merge"
```

Comment (zhaohuabing) [reply api/config/v1alpha1/envoyproxy_types.go:164]: I'm not sure whether there're any use cases that need a list of `Merge`.

**Comment by zhaohuabing** [comment]
@arkodg Should I add a notice in the release note so that users won't be surprised by this breaking change? If yes, do you know where to add this note?

**Comment by arkodg** [comment]
> @arkodg Should I add a notice in the release note so that users won't be surprised by this breaking change? If yes, do you know where to add this note?

added the release note label, hopefully this reminds the release manager to add this note, we can work on proactively adding release notes too, maybe in the PR description that the release manager can copy into the release notes

File: api/config/v1alpha1/envoyproxy_types.go (hunk 2 above)
Comment (arkodg) [reply api/config/v1alpha1/envoyproxy_types.go:164]: cool, yah, most likely not needed

File: internal/infrastructure/kubernetes/proxy/resource.go [hunk 3]
```diff
@@ -124,18 +125,29 @@ func expectedProxyContainers(infra *ir.ProxyInfra, deploymentConfig *egcfgv1a1.K
...
 	if infra.Config != nil &&
 		infra.Config.Spec.Bootstrap != nil {
-		bootstrapConfigurations = *infra.Config.Spec.Bootstrap
-	} else {
-		var err error
-		// Use the default Bootstrap
-		bootstrapConfigurations, err = bootstrap.GetRenderedBootstrapConfig(proxyMetrics)
-		if err != nil {
-			return nil, err
+		userBootstrapConfigurations := infra.Config.Spec.Bootstrap.Value
```
Comment (arkodg) [inline internal/infrastructure/kubernetes/proxy/resource.go:139]: seeing this style used in many places, can we add a `util` in `internal/xds/bootstrap` to improve reuse ?

File: internal/utils/yaml/yaml.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+import (
+	"reflect"
+
+	"sigs.k8s.io/yaml"
+)
+
+// MergeYAML merges two yaml files. The second yaml file will override the first one if the same key exists.
+// This method can add or override a value within a map, or add a new value to a list.
+// Please note that this method can't override a value within a list.
+func MergeYAML(base, override string) (string, error) {
```
Comment (arkodg) [inline internal/utils/yaml/yaml.go:17]: can we use https://github.com/evanphx/json-patch instead, i.e. convert the YAMLs into JSON using `sigs.k8s.io/yaml` and then use `jsonpatch.MergePatch` and convert it back to YAML ?

**Comment by arkodg** [comment]
can we also update the user docs https://gateway.envoyproxy.io/v0.5.0/user/customize-envoyproxy.html#customize-envoyproxy-bootstrap-config in this PR or as a follow up, thanks !

File: api/config/v1alpha1/envoyproxy_types.go [hunk 5]
```diff
@@ -56,7 +56,7 @@ type EnvoyProxySpec struct {
 	// Bootstrap configuration used. You can edit this configuration, and rerun `egctl x translate` to ensure there are no validation errors.
 	//
 	// +optional
-	Bootstrap *string `json:"bootstrap,omitempty"`
```
Comment (zirain) [inline api/config/v1alpha1/envoyproxy_types.go:59]: this's a breaking change?

File: internal/utils/yaml/yaml.go (hunk 4 above)
Comment (zhaohuabing) [reply internal/utils/yaml/yaml.go:17]: `jsonpatch.MergePatch` can't be used to merge two JSON strings, and there are no other methods to merge JSON strings either.

See https://github.com/evanphx/json-patch#create-and-apply-a-merge-patch

File: api/config/v1alpha1/envoyproxy_types.go (hunk 5 above)
Comment (zhaohuabing) [reply api/config/v1alpha1/envoyproxy_types.go:59]: Yes, it's a breaking change. We can make it non-breaking by leaving the current `Booststrap` as a string and adding a new field, but I can't think of a good name for the new field. Do you have any idea?

Comment (zirain) [reply api/config/v1alpha1/envoyproxy_types.go:59]: I'm OK with this, but must highlight in release-notes.

Comment (arkodg) [reply api/config/v1alpha1/envoyproxy_types.go:59]: yes, added the release-notes tag

**Review by zirain — DISMISSED** [review]
LGTM, thanks @zhaohuabing !

**Review by arkodg — APPROVED** [review]
LGTM thanks !

**Pull request merged**